services/src/main/java/org/keycloak/services/util/DPoPUtil.java

```diff
@@ -0,0 +1,327 @@
+        } catch (VerificationException ex) {
+            throw new VerificationException("Token verification failure", ex);
+        }
+    }
+
+    public static void bindToken(AccessToken token, DPoP dPoP) {
+        AccessToken.Confirmation confirmation = token.getConfirmation();
+
+        if (confirmation == null) {
+            confirmation = new AccessToken.Confirmation();
+            token.setConfirmation(confirmation);
+        }
+
+        confirmation.setKeyThumbprint(dPoP.getThumbprint());
+    }
+
+    private static class DPoPClaimsCheck implements TokenVerifier.Predicate<DPoP> {
+
+        static final TokenVerifier.Predicate<DPoP> INSTANCE = new DPoPClaimsCheck();
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            Long iat = t.getIat();
+            String jti = t.getId(), htu = t.getHttpUri(), htm = t.getHttpMethod();
+
+            if (iat != null &&
+                jti != null && !jti.trim().equals("") &&
+                htm != null && !htm.trim().equals("") &&
+                htu != null && !htu.trim().equals("")) {
+                return true;
+            } else {
+                throw new DPoPVerificationException(t, "DPoP mandatory claims are missing");
+            }
+        }
+
+    }
+
+    private static class DPoPHTTPCheck implements TokenVerifier.Predicate<DPoP> {
+
+        private final HttpRequest request;
+        private final UriInfo uri;
+
+        DPoPHTTPCheck(HttpRequest request, UriInfo uri) {
+            this.request = request;
+            this.uri = uri;
+        }
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            try {
+                if (new URI(t.getHttpUri()).equals(uri.getAbsolutePath()) &&
+                    request.getHttpMethod().equals(t.getHttpMethod())) {
+                    return true;
+                } else {
+                    throw new DPoPVerificationException(t, "DPoP HTTP method/URL mismatch");
+                }
+            } catch (URISyntaxException ex) {
+                throw new DPoPVerificationException(t, "Malformed HTTP URL in DPoP proof");
+            }
+        }
+
+    }
+
+    private static class DPoPReplayCheck implements TokenVerifier.Predicate<DPoP> {
+
+        private final KeycloakSession session;
+        private final int lifetime;
+
+        public DPoPReplayCheck(KeycloakSession session, ClientModel client) {
+            this.session = session;
+            OIDCAdvancedConfigWrapper config = OIDCAdvancedConfigWrapper.fromClientModel(client);
+            this.lifetime = config.getDPoPProofLifetime();
+        }
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            SingleUseTokenStoreProvider singleUseCache = session.getProvider(SingleUseTokenStoreProvider.class);
+            byte[] hash = HashUtils.hash("SHA1", (t.getId() + "\n" + t.getHttpUri()).getBytes());
```

Comment:
As [DPoP's specification](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop#section-4.2) mentions, it might be sufficient to use only "jti". Could you tell me why you include "htu"?

Reply:
The spec also [says](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop#section-10.1):

>    Servers SHOULD store, **in the context of the request URI**, the `jti`
   value of each DPoP proof for the time window in which the respective
   DPoP proof JWT would be accepted and decline HTTP requests **to the
   same URI** for which the `jti` value has been seen before.

From this, I conclude that the `jti` values should be URI-scoped. The simplest way to enforce one-time use of a URI-scoped `jti` is to incorporate the URI representation into the single use cache key. Hashing is used to guard against any unexpected characters in input strings, but maybe I'm somewhat overcautious here.

Reply:
Just in case, `DPoPReplayCheck` is applied after `DPoPHTTPCheck`, so it is guaranteed that `htu` equal the actual request URI.

Reply:
Thank you for clarification. I agree with you.